[BPF] support for ipv4 options

tomastigera opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

## Related issues/PRs

<!-- If appropriate, include a link to the issue this fixes.
fixes <ISSUE LINK>

If appropriate, add links to any number of PRs documented by this PR
documents <PR LINK>
-->

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

tomastigera committed:
[WIP] ipv4 opts test

tomastigera committed:
[BENCH]

tomastigera committed:
WIP

tomastigera committed:
sort of works

tomastigera committed:
parse icmp related

tomastigera committed:
full benchmark

tomastigera committed:
split nat and split off icmp related natting WIP

tomastigera committed:
[BPF] finished icmp

tomastigera committed:
[XDP] compiles

tomastigera committed:
gofmt

tomastigera committed:
[XDP] no ipv4 options

tomastigera committed:
[BPF] skip ipv6 parsing tests

tomastigera committed:
[BPF] load calico_tc_skb_icmp_inner_nat progs

tomastigera committed:
fix TestAttach

tomastigera committed:
[XDP] use bpf_xdp_load_bytes where available (5.18+)

tomastigera committed:
clean up

tomastigera committed:
[LIBBPF] fix build

tomastigera committed:
lock in bpf/mock/map

What is protected by the syscalls in the real workld neeeds to be
protected by a lock when mocking.

fasaxc commented on felix/bpf-gpl/globals.h:
```diff
@@ -16,11 +16,16 @@ struct cali_tc_globals {
 	__u8 iface_name[16];
 	__u32 log_filter_jmp;
 	__u32 jumps[32];
+	/* Needs to be 32bit aligned as it is followed by scratch area for
```
should we make the scratch area explicit here `__u32 scratch[xxx]`  Just thinking that removes a special case for this map and ensures the alignment?

tomastigera replied:
yup, made it explicit

fasaxc commented on felix/bpf-gpl/skb.h:
```diff
@@ -70,7 +70,8 @@ static CALI_BPF_INLINE long skb_iphdr_offset(struct cali_tc_ctx *ctx)
 		// Ingress on an IPIP tunnel: skb is [ether|outer IP|inner IP|payload]
 		// TODO: we need to consider different types of IPIP tunnels like 4in6 or 6in4
-		return sizeof(struct ethhdr) + ctx->ipheader_len;
+		// XXX no support for ip options in ipip heder
```
```suggestion
		// XXX no support for ip options in ipip header
```

fasaxc commented on felix/bpf-gpl/tc.c:
```diff
@@ -1042,8 +740,30 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
 
-		res |= bpf_l3_csum_replace(skb, l3_csum_off, state->ip_src, state->ct_result.nat_sip, 4);
-		res |= bpf_l3_csum_replace(skb, l3_csum_off, state->ip_dst, state->post_nat_ip_dst, 4);
+		if (!in_place) {
```
Any value in wrapping this up as a utility function, pops up in a few places (but maybe calling a utility function would be almost as verbose...

tomastigera replied:
yeah, thought about it a long time ago, but not worth it.

fasaxc reviewed (commented):
Feels a lot less scary after you talked me through the refactor.

I'm a little wary of copying the data out and into the packets again; seems like a bit of an overhead but perhaps that can be factored back out again.  Maybe we should just move wholesale to the bpf-helper-based load/store functions and abandon direct access.

coutinhop committed:
Update renamed windows AMIs

Windows AMIs were renamed from 'Windows_Server-*-English-Core-ContainersLatest' to 'Windows_Server-*-English-Core-Base', update them accordingly.

Also remove mentions to unsupported windows versions (e.g 1903, 1909, 2004, 20H2).

coutinhop committed:
Install docker because AMI doesn't have it anymore

mgleung committed:
Merge pull request #7928 from coutinhop/pedro-fix-win-ami

Update renamed windows AMIs

tomastigera committed:
[BPF] use explicit scratch pointer in state

Makes it always aligned and easy to reach.

tomastigera committed:
[BPF] fix typo

tomastigera committed:
[XDP] scratch comment

sridhartigera committed:
Upgrade libbpf 1.2.2 (#7922)

* Upgrade libbpf 1.2.2

sridhartigera committed:
Fix jump map key, value types (#7936)
